course5/homework6-try2/homework06.cc

```diff
@@ -0,0 +1,139 @@
+#include "mystr.h"
+#include <cstdio>
+#include <iostream>
+#include <cassert>
+#include <cstring>
+
+void test_operators();
+void test_basics();
+
+using namespace mystring;
+
+template <typename T>
+void test_basics()
+{
+	std::cout << "====== Testing basics ======" << std::endl;
+
+	T empty("");
+	T empty_too; 
+	std::cout << "created empty string" << std::endl;
+	assert(empty.length() == 0); 
+	assert(empty == empty_too); 
+	assert(empty == ""); 
+
+	// 2 empties should also be able to be non-compared
+	if (empty != empty_too)
+		assert(false);
+	std::cout << "creating full string with different operator" << std::endl;
+	T str("Hello world");
+	assert(str.length() == 11); 
+	assert(str ==  "Hello world" && str != "Hello worl"); 
+
+	std::cout << "Created empty string str='"<< empty << "' and full string str='" << str  << "'" << std::endl;
+
+	T str2 = str; // copy constructor
+	std::cout << "Created str2 from str= "<<  str2 << std::endl; 
+	assert(str == str2);
+
+	T str3;
+	str3 = "Haleluja"; // copy assignment with char*
+
+	std::cout << "str3 created and assigned by copy. SHould be Haleluja: "<<  str3 << std::endl; 
+	assert(str3 == "Haleluja");
+
+	str3 = str2; // copy assignment with T
+	std::cout << "str3. should be back to str2 (Hello World): "<<  str3 << std::endl; 
+	assert(str3== str2);
+	
+	// Try corner cases
+	empty = ""; // empty
+	T non_empty = "Non empty";
+	empty = non_empty; // Test assign non empty to empty
+	assert (empty == non_empty);
+
+	empty =  ""; 
+	T empty2 ; 
+	empty2 = empty; // assign empty to empty 
+	assert (empty2 == empty); 
+
+	non_empty = "full";
+	empty = non_empty; // assign non empty to empty
+	assert(empty == non_empty); 
+
+	std::cout << "~~~~~ Finished Testing basics " << std::endl;
+}
+
+template <typename T> 
+void test_operators()
+{
+	char test_str[] = "The Quick Brown Fox Jumps Over The Lazy Dog"; 
+	std::cout << "====== Testing operators ======" << std::endl;
+	T str(test_str);
+	T str2 = std::move(str);
+	T str3 = "Test string";
```

Comment:
here you can test if ```str``` is empty after the content has been stolen by ```str2```. The test may not work with std::string, thou, as one should make any assumption about the state of moved-from object.